src/Accelerators/NNPA/Conversion/ONNXToZHigh/RewriteONNXForZHigh.cpp

```diff
@@ -388,6 +388,60 @@ class SplitLargeMatMulPattern : public OpRewritePattern<ONNXMatMulOp> {
   }
 };
 
+/// This pattern is to replace `C = add/sub(A, B)` by `A` when B is a zero
+/// defined by ConstantOfShape and C's shape is the same as A's shape.
```

Comment:
I think all your rewrites might become a little bit simpler if you start by rewriting/canonicalizing `ConstantOfShape(value=x,shape=S)` to `Exand(Constant(value=x,shape=[]),shape=S)`

as @AlexandreEichenberger mentioned, here we could have a simple rule to rewrite `Add/Sub(A,Expand(B,S))` to `Add/Sub(A,B)` when shape(A)==S (and generalize it to any op that broadcasts it's operands)

then the const prop rules Add/SubZerosOnRhs would finish your rewrite here if you implement the const prop rules TODO to follow them with Expand

I think the other patterns
ReplaceBinaryOpByConstantOfShapePattern and
ReplaceUnsqueezeOfConstantOfShapeRewritePattern
also become a little bit simpler and more general if you define them for Expand after rewriting ConstantOfShape to Constant and Expand

I'll try to take a closer look at the whole PR tonight

Reply:
Thanks for the detailed explanation, I know understand how using the `constantOfShape` benefits from being expressed as a `expand(const(...), shape= )`.

In by original review, I was eager to see if we can do the constant propagation can happen when all constants are known but for the shape. Is there a way to sink the computations inside the extend?
```
add(const 1, expand(const 2, shape)) -> expand(add(const 1, const 2), shape) -> expand(const 3, shape)
```

Reply:
I like the idea of using `Exand(Constant(value=x,shape=[]),shape=S)` because we can reuse the constant propagation, and it also means we should call the constant propagation after each op rewriting. So for a long chain of operations, we should call `onnx-rewrite -> constprop -> onnx-rewrite -> ...` repeatedly, and I am afraid that it would be slow when we add more rules to `onnx-rewrite` in the future. Meanwhile, this patch eagerly folds constants to have a new ConstantOfShape, so that all ConstantOfShape-related ops can be folded in a single pass.

Reply:
> for a long chain of operations, we should call `onnx-rewrite -> constprop -> onnx-rewrite -> ...` repeatedly, and I am afraid that it would be slow when we add more rules to `onnx-rewrite` in the future

we could do onnx-rewrite together with constprop in a hybrid pass

I had always planned to include constprop in the hybrid pass, see the implementation in PR #2494

Reply:
It looks to me you gather all patterns to put them into a single pass HybridPass, so does it mean patterns in `onnx-rewrite` will be added to HybridPass too? and in the future all patterns related to onnx ops will go to HybridPass?

Reply:
> It looks to me you gather all patterns to put them into a single pass HybridPass, so does it mean patterns in `onnx-rewrite` will be added to HybridPass too? and in the future all patterns related to onnx ops will go to HybridPass?

yes, we could add onnx-rewrite to hybrid pass

you could also make another hybrid pass, e.g. with just onnx-rewrite and constprop, if that works best to achieve your goals

Reply:
How about using `ONNXOpTransformPass`? I am not clear which of `ONNXOpTransformPass` or `HybridPass` is used for which situation?

Reply:
> Is there a way to sink the computations inside the extend?
> `add(const 1, expand(const 2, shape)) -> expand(add(const 1, const 2), shape) -> expand(const 3, shape)`

I think this could be accomplished with the rule:
`Add(X,Expand(Y,shape=S)) -> Expand(Add(X,Y),shape=S)` when `X` has scalar shape (or any shape that is known to broadcast to Y's shape, but that's tricky to check more generally if Y has dynamic shape)

it can be generalized to any operand order and any broadcastable op with 2 or more operands

Reply:
fwiw, ReplaceUnsqueezeOfConstantOfShapeRewritePattern could maybe also be reformulated to act on Expand instead of ConstantOfShape so it could apply to Expand in other situations than just those that arise from ConstantOfShape

Reply:
> How about using `ONNXOpTransformPass`? I am not clear which of `ONNXOpTransformPass` or `HybridPass` is used for which situation?

I was hoping that hybrid pass would replace ONNXOpTransformPass

it should be more efficient because it's designed do everything in a single pass

it's also potentially simpler to understand the hybrid pass at a high level as just doing all the included sets of patterns (shape inference, canonicalization, constprop) at once compared to understanding the logic in ONNXOpTransformPass (iteration until convergence, and a sequence of passes with shape inference in between each)